### PR TITLE
Fix `CanvasItem.draw_rect` function with `filled = false`

### LINKED ISSUE #35384: draw_rect is drawn at wrong position. Also size and shape is wrong if filled parapeter is false.
**Godot version:**
3.2 RC 2 Open GLES 3

**OS/device including version:**
Windows 10
Nvidia GTX 1650

**Issue description:**
If filled is set to false, the bottom left pixel appears to be missing, but this is not correct. The size is actually one pixel too wide on both axes. Size and shape works as expected if filled is true. Also all rectangles appear to be drawn at wrong positions. See pictures below.
Related to #32279

Edit: problem is probably related to drawing lines: Drawing a 3 pixel line only shows 2 pixels, omitting the first one. Also horisontal lines are offset 1 pixel down on the y axis. Would probably be a good idea to fix lines first, if rectangles depend on them?

Blue and green are rectangles of 1 by 1 pixel size.
Red and blue are rectangles of 4 by 4 pixel size.
Green and blue have filled = false
Yellow and red have filled = true
Godot_v3 2-rc2_win64_7h1IAE1juE

filled = false drawn at Vector(1,1) - Shows at position (1,2)
Godot_v3 2-rc2_win64_qL4B1N6vcg

filled = true drawn at Vector(1,1) - Shows at position (2,2)
Godot_v3 2-rc2_win64_TXjYe5qXlT

**Steps to reproduce:**
Draw a rect with filled, and another without filled at the same size. Compare.

**Minimal reproduction project:**
[Draw_rect.zip](https://github.com/godotengine/godot/files/4088225/Draw_rect.zip)

### LINKED ISSUE #58775: A top corner pixel is missing in non-filled `CanvasItem.draw_rect()`
### Godot version

v4.0.alpha.custom_build [013efeef0]

### System information

Windows 10, Intel Core i5 7200U CPU, Vulkan

### Issue description

When using `CanvasItem.draw_rect()` with `fill = false` and `width = 1`, one of the top corner pixels are missing. From my observations, the top-left corner is missing when the pixel position is an integer and the top-right corner is missing when the pixel position has a decimal.

This also affects some areas of the editor, such as the 2D rectangle selection and the gradient editor:
image image

### Steps to reproduce

Each of these 3 options exhibits this bug:
a. Open the 2D editor and start a rectangle selection
b. Create a Gradient resource and add some points to the gradient editor
c. Attach a script to a CanvasItem-inherited node, and use `draw_rect()` with `fill = false` in the `_draw()` function.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by dalexeev**
Closes #35384.

---

Special thanks to @clayjohn and @kleonc for [the explanations](https://github.com/godotengine/godot/issues/41023).

### CODE REVIEW

**Comment by aaronfranke** [comment]
@dalexeev Is this still desired? If so, it needs to be [rebased](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) and tested on the latest master branch. While there are no conflicts, rebasing is important so that reviewers can easily test the PR.

**Comment by dalexeev** [comment]
R<sup>E<sup>B</sup>A</sup>S<sub>E<sub>D</sub>&nbsp;</sub>A<sup>N<sup>D</sup>&nbsp;</sup>R<sub>E<sub>A</sub>D</sub>Y<sup>&nbsp;<sup>F</sup>O</sup>R<sub>&nbsp;<sub>R</sub>E</sub>V<sup>I<sup>E</sup>W</sup>

File: scene/main/canvas_item.cpp [hunk 1]
```diff
@@ -784,40 +784,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+		int width = (int)Math::round(p_width);
+		real_t offset = 0.5 * width;
```
Comment (aaronfranke) [inline scene/main/canvas_item.cpp:795]: Nitpicking, but we can slightly optimize this. Multiplying an integer by 0.5 or by 0.5f should have the same exact behavior as long as this integer is less than 16777216, so we don't need the extra precision.

```suggestion
		real_t offset = 0.5f * width;
```

Comment (dalexeev) [reply scene/main/canvas_item.cpp]: Done.

File: scene/main/canvas_item.cpp [hunk 2]
```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
+		Size2 size = p_rect.size;
+		int width = (int)Math::round(p_width);
+		real_t offset = 0.5f * width;
+
+		if (width % 2) {
+			pos += Size2(0.5, 0.5);
+			size -= Size2(1, 1);
 		}
 
+		// top line
```
Comment (aaronfranke) [inline scene/main/canvas_item.cpp:527]:
```suggestion
		// Top line.
```

File: scene/main/canvas_item.cpp [hunk 3]
```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
 		RenderingServer::get_singleton()->canvas_item_add_line(
 				canvas_item,
-				p_rect.position + Size2(-offset, 0),
-				p_rect.position + Size2(p_rect.size.width + offset, 0),
+				pos + Size2(-offset, 0),
+				pos + Size2(-offset + size.width, 0),
 				p_color,
-				p_width);
+				width);
+		// right line
```
Comment (aaronfranke) [inline scene/main/canvas_item.cpp:534]:
```suggestion
		// Right line.
```

File: scene/main/canvas_item.cpp [hunk 4]
```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
 		RenderingServer::get_singleton()->canvas_item_add_line(
 				canvas_item,
-				p_rect.position + Size2(p_rect.size.width, offset),
-				p_rect.position + Size2(p_rect.size.width, p_rect.size.height - offset),
+				pos + Size2(size.width, -offset),
+				pos + Size2(size.width, -offset + size.height),
 				p_color,
-				p_width);
+				width);
+		// bottom line
```
Comment (aaronfranke) [inline scene/main/canvas_item.cpp:541]:
```suggestion
		// Bottom line.
```

File: scene/main/canvas_item.cpp [hunk 5]
```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
 		RenderingServer::get_singleton()->canvas_item_add_line(
 				canvas_item,
-				p_rect.position + Size2(p_rect.size.width + offset, p_rect.size.height),
-				p_rect.position + Size2(-offset, p_rect.size.height),
+				pos + Size2(offset + size.width, size.height),
+				pos + Size2(offset, size.height),
 				p_color,
-				p_width);
+				width);
+		// left line
```
Comment (aaronfranke) [inline scene/main/canvas_item.cpp:548]:
```suggestion
		// Left line.
```

**Review by aaronfranke — COMMENTED** [review]
Using the code from #35384 on this PR, if I have the base window size set to be small and let Godot stretch the Window to make things more visible, it looks correct:



However, if I scale up the Node2D instead of stretching the window, it doesn't look quite right:



I'm not super familiar with this function so I'm not sure if it's supposed to draw "one pixel thick" on the screen or relative to the Node2D, but this output is definitely not either.

Also, in addition to the blue rectangle not looking correct, the green one is drawn using filled set to false, but a size of one, but it looks like it's filled. If I change it to a size of not 1 it looks similarly broken to how the blue one is broken.

**Comment by dalexeev** [comment]
Thanks for the review. It looks like the bug in the `draw_line` function, it does not take into account scaling for 1 pixel lines, but does take into account other transformations.

Green and yellow lines are 1 pixel thick, blue and pink lines are 2 pixel thick. Node2D is scaled.



If I set the border thickness to 2 pixels, then the rectangle is drawn correctly:



I'll take a look at the draw_line function, but I'm not sure if I can fix this error.

File: scene/main/canvas_item.cpp [hunk 6]
```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+		int width = (int)Math::round(p_width);
+		real_t offset = 0.5f * width;
+
+		if (width % 2) {
+			pos += Size2(0.5, 0.5);
+			size -= Size2(1, 1);
 		}
```
Comment (dalexeev) [inline scene/main/canvas_item.cpp:525]: Actually, I would like someone to review this part. I'm not sure if this is the correct solution. Especially considering `CanvasItem` and/or `Viewport` scaling.

In fact, the `p_width` type needs to be replaced from `float` to `int`, since it is rounded. I just don't know how else to make the line appear exactly the width that is specified. Therefore, I take into account the width parity, in order to snap the lines to the pixel grid.

**Comment by dalexeev** [comment]
Also I would like to know which option is better.

**Comment by dalexeev** [comment]
I fixed the bug, but the solution is quite hacky. The main drawback is `p_width` rounding.

Also, there is a hack to check the width is 1 due to #51978:

https://github.com/godotengine/godot/blob/0056acf46fc88757cae9d9f6fe9805f0eec1cd09/servers/rendering/renderer_canvas_cull.cpp#L582-L602

In general, with integer scaling factors, everything is displayed correctly. For non-integer scales, lines can be drawn with different widths due to inaccurate mapping to the pixel grid. But I'm not sure if this problem can be solved at all.

With the current behavior, errors occur even at integer scales and linewidths, especially if the color is translucent.

**Comment by dalexeev** [comment]
> Does this fix #58775?

Yes.

```gdscript
draw_rect(Rect2(2, 2, 6, 6), Color.RED)
draw_rect(Rect2(2, 2, 6, 6), Color('yellow', 0.5), false)

draw_rect(Rect2(2, 12, 6, 6), Color.RED)
draw_rect(Rect2(2, 12, 6, 6), Color('yellow', 0.5), false, 2)

draw_rect(Rect2(2, 22, 6, 6), Color.RED)
draw_rect(Rect2(2, 22, 6, 6), Color('yellow', 0.5), false, 3)
```

**Comment by MewPurPur** [comment]
Does this fix the bug outlined below on the same issue, with corners being cut off when width is negative, and being repeated twice when height is positive?

On a personal branch I fixed it by getting x and y offsets from multiplying offset by the rect sizes' signs. But this branch modifies the logic further so this might not work.

**Comment by dalexeev** [comment]
> Also, does this fix the bug outlined below on the same issue, with corners being cut off when width is negative, and being repeated twice when height is negative?

I was unable to reproduce the bug for `rect.x < 0` (as stated in that comment) neither in beta 7 nor in this PR. Negative width/height is handled incorrectly in both cases, but at least there is a message "Rect2 size is negative, this is not supported. Use Rect2.abs() to get a Rect2 with a positive size".

**Comment by MewPurPur** [comment]
Huh, weird. Here's me -

```gdscript
@tool
extends Node2D

func _draw() -> void:
	draw_rect(Rect2(40, 40, -33, -33), Color(0, 1, 0, 0.5), false, 4.0)
	draw_rect(Rect2(56, 40, 33, -33), Color(1, 1, 0, 0.5), false, 4.0)
	draw_rect(Rect2(40, 56, -33, 33), Color(1, 0, 0, 0.5), false, 4.0)
	draw_rect(Rect2(56, 56, 33, 33), Color(0, 1, 0, 0.5), false, 4.0)

func _process(delta: float) -> void:
	queue_redraw()
```

Screenshot from 2022-11-30 17-55-29

No warnings thrown. Or maybe I misunderstood them.

**Comment by dalexeev** [comment]
Oh sorry. This warning only appears for `filled = true` for some reason. We need to either add the same warning for `filled = false` or take into account possible negative values. But in any case, a maintainer review is especially needed here, since this code is full of hacks.

File: scene/main/canvas_item.cpp [hunk 7]
```diff
@@ -525,38 +525,52 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
 
 		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect, p_color);
+	} else if (p_width >= p_rect.size.width || p_width >= p_rect.size.height) {
+		Size2 pos = p_rect.position - Size2(0.5 * p_width, 0.5 * p_width);
+		Size2 size = p_rect.size + Size2(p_width, p_width);
+		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, Rect2(pos, size), p_color);
```
Comment (kleonc) [inline scene/main/canvas_item.cpp:531]: This part is great. :+1:

**Comment by dalexeev** [comment]
Thanks, your script is very helpful. I removed the unnecessary rounding in the case of non-integer `p_width`s, but left the grid snapping for odd `p_width`s and the hack with `p_width <= 1.001` (I believe this is more correct in a practical sense, but I will listen to your opinion if you think differently). Except for this, the rendering is exactly the same as what your script draws (I tested negative values and scaling as well).

However, I'm not entirely sure that the behavior is correct for a negative rectangle width, although it matches your script:

File: scene/main/canvas_item.cpp [hunk 8]
```diff
@@ -557,38 +557,53 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+
+		// Snap to pixel grid for odd `p_width`.
+		if (Math::is_equal_approx(Math::fmod(p_width, 2), 1)) {
+			pos += Size2(0.5, 0.5);
+			size -= Size2(1, 1);
```
Comment (kleonc) [inline scene/main/canvas_item.cpp:571]: Again, I don't get what's the logic behind that kind of strange cases. It does break things when the condition is satisfied.

<details>
<summary>Example:</summary>

ArXQ3IhBR5
godot windows editor x86_64_2yO3J9qvS4
8g6mbuTXmT

</details>

What issue is it attempting to solve?

**Comment by kleonc** [comment]
> but left the grid snapping for odd `p_width`s and the hack with `p_width <= 1.001` (I believe this is more correct in a practical sense, but I will listen to your opinion if you think differently)

Honestly, I kinda ignored how you handle this case. It feels like it needs to finally be addressed properly by allowing the user to choose how the width should be interpreted (then you won't need any special casing in here). Personally I don't think it should ever auto-decide to use line-primitive over the quads. Just because I'm rendering a line with 0.01 width doesn't mean that it should be drawn as an actual line primitive on the GPU. I can have a camera zoomed in close and thus the resulting line can end up visually quite thick. And if I want a thin line regardless of the zoom etc. I should also be able to choose to draw it as such.

Not sure what's the opinion of the rendering maintainers on how this should/can be handled, what needs to be taken into account etc. But yeah, I think it definitely needs to be solved for 4.0.

A possible easy solution would be to make a negative widths be interpreted as "I want a line primitive", positive widths always as quads, and changing the default values for parameters to -1 so it would still not draw quads by default. But again, not sure (yet) if it's fine/acceptable solution.
 
> However, I'm not entirely sure that the behavior is correct for a negative rectangle width although it matches your script

I haven't though about negative values at all when writing that script. So not sure what the behavior in there is and what it should be. As far as I remember in case of rects and negative sizes there are some inconsistencies, in some places they are kinda forbidden / throwing warnings/errors and in other they have well defined behavior (e.g. when drawing texture to a destination rect with negative size it results in flipping the texture on the given axis when it's being drawn).

**Comment by MewPurPur** [comment]
> Personally I don't think it should ever auto-decide to use line-primitive over the quads. Just because I'm rendering a line with 0.01 width doesn't mean that it should be drawn as an actual line primitive on the GPU.

It's its whole can of worms - there were like three different PRs trying to solve this, but then Chaosus lost interest. Now it's a compat-breaker.

**Comment by dalexeev** [comment]
Without this hack, I have no idea how to solve this: https://github.com/godotengine/godot/pull/41239#pullrequestreview-735543356.

File: scene/main/canvas_item.cpp (hunk 8 above)
Comment (dalexeev) [reply scene/main/canvas_item.cpp:571]: Maybe we could add an optional `pixel_perfect` parameter?

**Comment by kleonc** [comment]
> It's its whole can of worms - there were like three different PRs trying to solve this, but then Chaosus lost interest. Now it's a compat-breaker.

@MewPurPur The question is kinda what can of worms it is. If it's just about ensuring in hundreds of places in the code that a proper thing is used according to whatever change would be made then I'd say it's fine and doable. The problem would be if such changes would break something in not really fixable manner. However, not sure what could get broken in such way.
Compat-breaker or not, it feels like it needs to finally be addressed at some point, whether it's already too late for 4.0 or not (4.1? not sure). It's not good that I have to use `draw_mesh`/`draw_polygon`/etc. to draw a rectangle border the way I want just because the `draw_rect` is not capable of doing so. And it's not just the usability aspect, it can potentially also degrade the performance as it would likely break batching.

> Without this hack, I have no idea how to solve this: [#41239 (review)]

@dalexeev Do you by any chance have a 4.0 project to provide for testing this?

**Comment by MewPurPur** [comment]
@kleonc In that case, we'd need to modify some other functions too, like `draw_line()`

**Comment by dalexeev** [comment]
> Do you by any chance have a 4.0 project to provide for testing this?

I did not quite understand, this can be tested on any project. Lines that are one pixel wide stay thin when scaled, so it's not surprising that the rectangle splits into 4 lines.

изображение

File: scene/main/canvas_item.cpp (hunk 8 above)
Comment (dalexeev) [reply scene/main/canvas_item.cpp]: I changed my mind. This point should be explained in the documentation/tutorial, and then the user can adjust the position and size for odd widths when calling `draw_rect` if they wish.

File: scene/main/canvas_item.cpp [hunk 9]
```diff
@@ -557,37 +557,41 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		// Workaround for scaling issue. See `RendererCanvasCull::canvas_item_add_line()`.
+		if (p_width <= 1.001) {
+			p_width = 1.002;
 		}
```
Comment (kleonc) [inline scene/main/canvas_item.cpp:566]: @dalexeev I kinda understand where this change is coming from but I don't think it's acceptable to be merged. Besides it not really fixing the scaling issue (it indeed makes such rects be always scaled but they'd still end up being wider than desired), it also breaks all the cases where drawing a rect with line primitive was desired (e.g. probably most `draw_rect` calls within the built-in GUI controls for drawing unfilled rect with undefined line width (hence using the default 1.0)).

Comment (dalexeev) [reply scene/main/canvas_item.cpp:566]: If there was a way
choose the desired `draw_line` option (2 point primitive vs 4 point primitive), then this hack would not be needed.

For the "thin rectangle" variant (which doesn't change line thickness when scaled), we have to consider the CanvasItem's transform?

I think one possible solution to this dilemma is to separate the API, like so:

* `draw_line` - always draws a 4 point primitive.
* `draw_thin_line` - draws a 2 point primitive.

Comment (kleonc) [reply scene/main/canvas_item.cpp:566]: > If there was a way
> choose the desired `draw_line` option (2 point primitive vs 4 point primitive), then this hack would not be needed.

Indeed, that's the problem. I've already mentioned it in https://github.com/godotengine/godot/pull/41239#issuecomment-1335792985) (and suggested another possible solution). If going for separating the API (again, can't tell if that's the preffered way of solving it) then I think it'd make more sense to keep the same prefix in both methods for easier discoverability, e.g. something like `draw_line`+`draw_line_primitive` / `draw_line_quad`+`draw_line` / `draw_line_quad`+`draw_line_primitive` / etc.

>For the "thin rectangle" variant (which doesn't change line thickness when scaled), we have to consider the CanvasItem's transform?

If we'd want to keep auto-detecting when to use line primitive over the quad and make it be done only when the actual result is thin (instead of when the rect is locally thin like it's currently done) then yeah, all the transforms which will affect the final CanvasItem render would need to be taken into account. But I don't think it's easily feasible as the decision which primitive to use (line/quad) would need to be probably postponed to when the RenderingServer actually issues stuff to be rendered. And actually I think auto detection might be not wanted as it might be break batching. It seems that primitives with different point count (like lines and quads) are currently not batched together. Which means using always quads could be potentially more performant than sometimes using line primitives. So yeah, I still think just giving the choice for quad/line to the user is the way to go.

Comment (kleonc) [reply scene/main/canvas_item.cpp:566]:
```suggestion
		// Don't offset if line primitives instead of quads. See `RendererCanvasCull::canvas_item_add_line()`.
		if (p_width <= 1.001) {
			offset = 0.0;
		}
```

File: scene/main/canvas_item.cpp [hunk 10]
```diff
@@ -557,37 +557,41 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
...
 
+		real_t offset = 0.5f * p_width;
```
Comment (kleonc) [inline scene/main/canvas_item.cpp:568]:
```suggestion
		real_t offset = 0.5f * p_width;

		// Don't offset if line primitives instead of quads. See `RendererCanvasCull::canvas_item_add_line()`.
		if (p_width <= 1.001) {
			offset = 0.0;
		}
```

For reference: https://github.com/godotengine/godot/pull/69851#issuecomment-1345398511.

Comment (dalexeev) [reply scene/main/canvas_item.cpp:568]: The lines will overlap in the corners, which will be noticeable with a translucent color.



The only option to remove this hack is #69851 or something similar. But I don't know how to make it work for thin lines (`width < 0`) in an easy way. In any case, `offset` must not be 0.

File: scene/main/canvas_item.cpp (hunk 6 above)
Comment (dalexeev) [reply scene/main/canvas_item.cpp]: I removed this part.

**Comment by akien-mga** [comment]
Superseded by #69851.

**Comment by akien-mga** [comment]
Or maybe not and it needs to be rebased on top of #69851?

**Comment by dalexeev** [comment]
There are a couple of things in this PR that are not in #69851. I will rebase it soon.

**Comment by dalexeev** [comment]
Essentially this PR is now cosmetic, the main bug was fixed in #69851. For negative `width`s, #58775 remains, but there is now a note about it in the docs. @kleonc, is there anything else we can do to reduce the chance of missing corners?

**Comment by kleonc** [comment]
> Essentially this PR is now cosmetic, the main bug was fixed in #69851. For negative `width`s, #58775 remains, but there is now a note about it in the docs. @kleonc, is there anything else we can do to reduce the chance of missing corners?

I think #58775 is not really solvable in a general case. It's not just a matter of obtaining proper final positions of the line endpoints in the coordinate space of the viewport/framebuffer/render target (or whatever you call it) which would be needed to be able to apply a proper offset in that space (small final offset; regardless of the zoom etc.). What we'd also need to know to determine what such offset should be is the exact algorithm used by the Vulkan/OpenGL/etc. for rasterizing the line primitives. And that's problematic, because these can differ between implementations since specifications aren't super strict:
- [OpenGL 3.3 Basic Line Segment Rasterization](https://registry.khronos.org/OpenGL/specs/gl/glspec33.core.pdf#subsection.3.5.1) (for [4.6 it's pretty much the same](https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf#subsection.14.5.1)),
- [Vulkan 1.1 Line Segments Rasterization](https://registry.khronos.org/vulkan/specs/1.1/html/chap25.html#primsrast-lines).

So even assuming that rasterization is done according to the diamond exit rule (which is not guaranteed; but maybe it's actually the most common case?:thinking: no idea about that) we would still need to find a way to somehow obtain the final positions of the line endpoints within the fragments' coordinate space (I don't see how we could currently do it). And then we would still need to find a way of how to offset the vertices (probably branching into many different cases) to always produce satisfying fragments. E.g. take a look at the image [from this post](https://stackoverflow.com/a/10341863) and imagine all possible configurations of the vertices (colored dots); and yeah, the rect can end up being transformed in there (so take into consideration rotations or... skewing; yeah, good luck with that :smile:).

**TL;DR**: I don't see us solving it. Would love to be surprised by someone but I don't think it will happen.

For a non-general solution / hacky solution what maybe could be done is adding a method (are a bool param to the current `draw_rect`) which would be supposed to work properly only in a very specific conditions: when the given CanvasItem is not transformed so the offset we would apply to the vertices would actually be equivalent to the offset in the fragments' space. Still it wouldn't be guaranteed to always work because of the loose specifications mentioned previously. Anyway, I'm not convinced that trying to handle such edge case is a good idea.

---

Regarding #35384 I guess it's solved by combination of #69851 + this PR + https://github.com/godotengine/godot-docs/pull/6455?

**Review by kleonc — COMMENTED** [review]
There's a question whether we should handle rects with negative size properly? Currently (AFAICT, haven't tested) only filled ones would be drawn correctly (I mean that `rect` with negative size would be drawn the same as `rect.abs()`). Shouldn't be hard to handle them too.

**Comment by dalexeev** [comment]
> There's a question whether we should handle rects with negative size properly?

Fixed. `Rect2 rect = p_rect.abs();` is OK?

**Comment by dalexeev** [comment]
> I wonder if there is any specific reason to prefer one over the other.

I'm not sure since this PR started quite a while ago and these changes have carried over through all the rebases. I will remove them if they are unnecessary.

**Comment by kleonc** [comment]
> I'm not sure since this PR started quite a while ago and these changes have carried over through all the rebases. I will remove them if they are unnecessary.

To me they do seem like not changing anything (so leaving the changes should be fine too).

**Review by kleonc — APPROVED** [review]
LGTM. Would welcome @clayjohn's opinion on https://github.com/godotengine/godot/pull/41239#issuecomment-1385618227 before someone will merge this.

File: scene/main/canvas_item.cpp [hunk 11]
```diff
@@ -557,44 +557,47 @@ void CanvasItem::draw_multiline_colors(const Vector<Point2> &p_points, const Vec
...
 				p_width);
+		// Left line.
 		RenderingServer::get_singleton()->canvas_item_add_line(
 				canvas_item,
-				p_rect.position + Size2(0, p_rect.size.height - offset),
-				p_rect.position + Size2(0, offset),
+				rect.position + Size2(0, offset + rect.size.height),
+				rect.position + Size2(0, offset),
 				p_color,
 				p_width);
```
Comment (dalexeev) [inline scene/main/canvas_item.cpp:602]: Should `draw_line`s be replaced with `canvas_item_add_polyline` after #71679 is merged? Like here:

https://github.com/godotengine/godot/blob/14fdd28de9f1f9979cb6cffa4b715a16d406439d/scene/resources/rectangle_shape_2d.cpp#L74-L88

Comment (clayjohn) [reply scene/main/canvas_item.cpp:602]: I guess so. A polyline should be slightly faster than having multiple regular lines

**Comment by clayjohn** [comment]
> LGTM. Would welcome @clayjohn's opinion on [#41239 (comment)](https://github.com/godotengine/godot/pull/41239#issuecomment-1385618227) before someone will merge this.

I don't think there is a good reason to prefer one behaviour over the other.

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
